azure/defaults_test.go

```diff
@@ -24,11 +24,99 @@ import (
 	"sync"
 	"testing"
 
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
 	"github.com/Azure/go-autorest/autorest"
 	. "github.com/onsi/gomega"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
+func TestARMClientOptions(t *testing.T) {
+	g := NewWithT(t)
+
+	tests := []struct {
+		name          string
+		cloudName     string
+		expectedCloud cloud.Configuration
+		expectError   bool
+	}{
+		{
+			name:          "should return default client options if cloudName is empty",
+			cloudName:     "",
+			expectedCloud: cloud.Configuration{},
+		},
+		{
+			name:          "should return Azure public cloud client options",
+			cloudName:     PublicCloudName,
+			expectedCloud: cloud.AzurePublic,
+		},
+		{
+			name:          "should return Azure China cloud client options",
+			cloudName:     ChinaCloudName,
+			expectedCloud: cloud.AzureChina,
+		},
+		{
+			name:          "should return Azure government cloud client options",
+			cloudName:     USGovernmentCloudName,
+			expectedCloud: cloud.AzureGovernment,
+		},
+		{
+			name:        "should return error if cloudName is unrecognized",
+			cloudName:   "AzureUnrecognizedCloud",
+			expectError: true,
+		},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			opts, err := ARMClientOptions(tc.cloudName)
+			if tc.expectError {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(opts.Cloud).To(Equal(tc.expectedCloud))
+			g.Expect(opts.Retry.MaxRetries).To(BeNumerically("==", -1))
+			g.Expect(opts.PerCallPolicies).To(HaveLen(2))
+		})
+	}
+}
+
+func TestPerCallPolicies(t *testing.T) {
+	g := NewWithT(t)
+
+	corrID := "test-1234abcd-5678efgh"
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		g.Expect(r.Header.Get("User-Agent")).To(ContainSubstring("cluster-api-provider-azure/"))
+		g.Expect(r.Header.Get(string(tele.CorrIDKeyVal))).To(Equal(corrID))
+		fmt.Fprintf(w, "Hello, %s", r.Proto)
+	}))
+	defer server.Close()
+
+	opts, err := ARMClientOptions("")
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(opts.PerCallPolicies).To(HaveLen(2))
+	ctx := context.WithValue(context.Background(), tele.CorrIDKeyVal, tele.CorrID(corrID))
```

Comment:
Just curious, why do we need a custom context here?

Reply:
This is actually how the `correlationIDPolicy` works: it looks for a Microsoft "correlation ID" in the context of the request object. 

So here we're setting up what ends up as a request "decorator" for the test that verifies the behavior of the `correlationIDPolicy`.

In SDKv1 it's done this way for comparison:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/a0dcdea1b8168babca3fc8935345a8026f89d568/azure/defaults.go#L336-L339